File: go/301.go
func removeInvalidParentheses(s string) []string {
    var ret []string
    visited, queue, found := map[string]bool{s: true}, []string{s}, false
    for len(queue) > 0 {
        cur := queue[0]
        queue = queue[1:]
        if isValid(cur) {
            ret = append(ret, cur)
            found = true
        }
        if found {
            continue
        }
        for idx, c := range cur {
            if c != '(' && c != ')' {
                continue
            }
            temp := cur[:idx] + cur[idx+1:]
            if _, ok := visited[temp]; !ok {
                visited[temp] = true
                queue = append(queue, temp)
            }
        }
    }
    return ret
}

func isValid(str string) bool {
    cnt := 0
    for _, c := range str {
        if c == '(' {
            cnt++
        } else if c == ')' {
            cnt--
            if cnt < 0 {
                return false
            }
        }
    }
    return cnt == 0
}

File: cpp/301.cpp
class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        vector<string> ret;
        unordered_set<string> visited{{s}};
        queue<string> q{{s}};
        bool found = false;
        while (!q.empty()) {
            string temp = q.front();
            q.pop();
            if (isValid(temp)) {
                ret.push_back(temp);
                found = true;
            }
            if (found)
                continue;
            for (int size = temp.size(), i = 0; i < size; ++i) {
                if (temp[i] != '(' && temp[i] != ')')
                    continue;
                string tmp = temp.substr(0, i) + temp.substr(i+1);
                if (!visited.count(tmp)) {
                    q.push(tmp);
                    visited.insert(tmp);
                }
            }
        }
        return ret;
    }
    bool isValid(string t) {
        int cnt = 0, size = t.size();
        for (int i = 0; i < size; ++i) {
            if (t[i] == '(')
                ++cnt;
            else if (t[i] == ')' && --cnt < 0)
                return false;
        }
        return cnt == 0;
    }
};
